internal/uvm/scsi.go

```diff
@@ -26,67 +25,142 @@ var (
+	vm *UtilityVM
+	// path is the host path to the vhd that is mounted.
+	HostPath string
+	// path for the uvm
+	UvmPath string
+	// scsi controller
+	Controller int
+	// scsi logical unit number
+	Lun int32
+	// While most VHDs attached to SCSI are scratch spaces, in the case of LCOW
+	// when the size is over the size possible to attach to PMEM, we use SCSI for
+	// read-only layers. As RO layers are shared, we perform ref-counting.
+	isLayer  bool
+	refCount uint32
+}
+
+func (sm *SCSIMount) String() string {
+	scsiFormat := "{HostPath: %s, UvmPath: %s, isLayer: %t, refCount: %d, Controller: %d, Lun: %d}"
+	return fmt.Sprintf(scsiFormat, sm.HostPath, sm.UvmPath, sm.isLayer, sm.refCount, sm.Controller, sm.Lun)
+}
+
 // allocateSCSI finds the next available slot on the
 // SCSI controllers associated with a utility VM to use.
 // Lock must be held when calling this function
-func (uvm *UtilityVM) allocateSCSI(ctx context.Context, hostPath string, uvmPath string, isLayer bool) (int, int32, error) {
+func (uvm *UtilityVM) allocateSCSI(ctx context.Context, hostPath string, uvmPath string, isLayer bool) (*SCSIMount, error) {
 	for controller, luns := range uvm.scsiLocations {
 		for lun, si := range luns {
-			if si.hostPath == "" {
-				uvm.scsiLocations[controller][lun].hostPath = hostPath
-				uvm.scsiLocations[controller][lun].uvmPath = uvmPath
-				uvm.scsiLocations[controller][lun].isLayer = isLayer
-				uvm.scsiLocations[controller][lun].refCount = 1
-				log.G(ctx).WithFields(logrus.Fields{
-					"hostPath":   hostPath,
-					"uvmPath":    uvmPath,
-					"isLayer":    isLayer,
-					"refCount":   1,
-					"controller": controller,
-					"lun":        lun,
-				}).Debug("allocated SCSI location")
-				return controller, int32(lun), nil
+			// If si is nil, we have found an open slot so we allocate a new SCSIMount
+			if si == nil {
+				uvm.scsiLocations[controller][lun] = &SCSIMount{
+					HostPath:   hostPath,
+					UvmPath:    uvmPath,
+					isLayer:    isLayer,
+					refCount:   1,
+					Controller: controller,
+					Lun:        int32(lun),
+				}
+				log.G(ctx).Debug("allocated SCSI location: ", uvm.scsiLocations[controller][lun])
+				return uvm.scsiLocations[controller][lun], nil
 			}
 		}
 	}
-	return -1, -1, ErrNoAvailableLocation
+	return nil, ErrNoAvailableLocation
 }
 
-func (uvm *UtilityVM) deallocateSCSI(ctx context.Context, controller int, lun int32) {
+func (uvm *UtilityVM) deallocateSCSI(ctx context.Context, sm *SCSIMount) {
 	uvm.m.Lock()
 	defer uvm.m.Unlock()
-	si := uvm.scsiLocations[controller][lun]
-	if si.hostPath != "" {
-		log.G(ctx).WithFields(logrus.Fields{
-			"hostPath":   si.hostPath,
-			"uvmPath":    si.uvmPath,
-			"isLayer":    si.isLayer,
-			"refCount":   si.refCount,
-			"controller": controller,
-			"lun":        lun,
-		}).Debug("removed SCSI location")
-		uvm.scsiLocations[controller][lun] = scsiInfo{}
+	if sm != nil {
+		log.G(ctx).Debug("removed SCSI location: ", sm)
+		sm = nil
```

Comment:
This won't work. You're just setting the local variable `sm` to nil. You need to update the value in the `scsiLocations` array to be nil.